workspaces/ui/src/contexts/TestingDashboardContext.js

```diff
@@ -59,3 +86,66 @@ export function useEndpointPath(captureId, endpointId) {
 }
 
 export { queriesFromEvents } from '../services/TestingService';
+
+// Reducers
+// --------
+
+function initialServiceState() {
```

Comment:
have you checked out xstate? https://xstate.js.org/

Reply:
Never have! Pretty happy with the simplicity of this, but I can see how it might be useful it the amount of states and state transitions increased.

Reply:
I'd say this is right at the point where I'd want to switch over to state machines, only because we don't have it set up yet. So I wouldn't block this change, but if we had it set up I'd want to use it by default. I think even in the simplest case it's never worse / more complex.

Reply:
Must admit my personal point of friction lies quite a bit further down the road. As long as I can sequentially read over a function and see what's going on without massive if statements to only allow the right transitions, I've never really had trouble getting them to work correctly and robustly. Either way, for now I've left this as is, choosing to ship less code. Since it's completely unabstracted and pure code, the path for applying an abstraction on top of it when necessary is completely open.